### ISSUE TITLE
12$: wrong precedence: 'undefined' === typeof foo ? 3 : 4

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#306]
The following code has no errors, but quick-lint-js reports E057:
```js
 'undefined' === typeof foo ? 3 : 4;
```
It looks like precedence is the issue. quick-lint-js is happy with the following code:
```js
 'undefined' === (typeof foo) ? 3 : 4;
```

---

Possibly related: The following emits E020 but shouldn't:
```js
typeof true ? true : b = a
```

**Comment by singalhimanshu** [comment]
I claim this for-hire task. I expect payment after I complete this task. I will email the quick-lint-js team if I am assigned this task.

**Comment by singalhimanshu** [comment]
I think the issue is coming for here - https://github.com/quick-lint/quick-lint-js/blob/2ec61fbf9f28f8df7503d14e7fce81a0b58baac7/src/parse.cpp#L220-L223

```js
'undefined' === typeof foo ? 3 : 4; // use of undeclared variable: foo [E057]
//                     ^^^^^^^^^^^ this is considered as an expression_kind::conditional operator.
```
Whereas the expected behavior is that `foo` is considered as an `expression_kind::variable`, but currently it is `expression_kind::conditional` which later causes problem here - https://github.com/quick-lint/quick-lint-js/blob/385b32e0c373859805b30be2620a4a958ad00e3e/src/quick-lint-js/parse.h#L744-L748 
A possible solution can be making `binary_operators=false` here - https://github.com/quick-lint/quick-lint-js/blob/2ec61fbf9f28f8df7503d14e7fce81a0b58baac7/src/parse.cpp#L220-L223 but that won't fix it because a code like this will be interpreted incorrectly -
```js
typeof foo.bar; 
```

**Comment by strager** [comment]
> A possible solution can be making `binary_operators=false` here -
> 
> https://github.com/quick-lint/quick-lint-js/blob/2ec61fbf9f28f8df7503d14e7fce81a0b58baac7/src/parse.cpp#L220-L223
> but that won't fix it because a code like this will be interpreted incorrectly -
> 
> ```js
> typeof foo.bar; 
> ```

Sounds like we need a new precedence bool then.

(Our precedence parsing is crap. I'd welcome something better. ;P)

**Comment by singalhimanshu** [comment]
> Sounds like we need a new precedence bool then.

I don't understand how precedence will help here. Can you explain with this example? 
```js
'undefined' === typeof foo ? 3 : 4;
```

**Comment by strager** [comment]
> > Sounds like we need a new precedence bool then.
> 
> I don't understand how precedence will help here. Can you explain with this example?
> 
> ```js
> 'undefined' === typeof foo ? 3 : 4;
> ```

```js
'undefined' === typeof foo ? 3 : 4;
'undefined' === (typeof (foo ? 3 : 4)); // how quick-lint-js interprets it now
('undefined' === (typeof foo)) ? 3 : 4; // how quick-lint-js should interpret it
```

When parsing `typeof`'s operand, we should stop parsing at `?`.

**Comment by singalhimanshu** [comment]
> When parsing `typeof`'s operand, we should stop parsing at `?`.

Yeah that is what I have tried to implement in the PR now.
